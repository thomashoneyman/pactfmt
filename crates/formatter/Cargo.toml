[package]
name = "formatter"
version = "0.1.0"
edition = "2021"

[dependencies]
syntax = { path = "../syntax" }
pretty = "0.12.3"

[dev-dependencies]
insta = "1.41.1"
pretty_assertions = "1.4.0"
test_each_file = "0.3.4"
test-utils = { path = "../test-utils" }
