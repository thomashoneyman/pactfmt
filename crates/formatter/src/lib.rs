mod format_doc;
mod format_tree;
mod lower_cst;

use logos::Logos;
use pretty::RcAllocator;
use syntax::{lexer_old, parser};

use crate::format_doc::FormatDoc;
use crate::lower_cst::lower_toplevel;

pub fn format_source(input: &str, width: usize) -> String {
    let lexed = lexer_old::Token::lexer(input);
    let lexed_ok: Vec<_> = lexed.filter_map(|token| token.ok()).collect();
    let parsed = parser::parse(&mut lexed_ok.as_slice()).expect("failed to parse");
    let lowered = parsed.into_iter().map(lower_toplevel).collect::<Vec<_>>();
    let format_doc = lowered
        .iter()
        .fold(FormatDoc::nil(&RcAllocator), |acc, fst| {
            acc.append(fst.format(&RcAllocator))
        });
    format_doc.pretty(width)
}

pub fn check_source(input: &str) -> bool {
    let lexed = lexer_old::Token::lexer(input);
    let lexed_ok: Vec<_> = lexed.filter_map(|token| token.ok()).collect();
    parser::parse(&mut lexed_ok.as_slice()).is_ok()
}
