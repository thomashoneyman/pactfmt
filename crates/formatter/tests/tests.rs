use formatter::format_source;
use test_each_file::test_each_file;

test_each_file! { in "crates/formatter/tests/fixtures" => |content: &str| {
    insta::assert_snapshot!(format_source(content, 40));
}}
