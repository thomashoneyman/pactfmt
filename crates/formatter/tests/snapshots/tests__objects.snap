---
source: crates/formatter/tests/tests.rs
expression: "format_source(content, 40)"
---

